### PR TITLE
Make oak_restricted_kernel_wrapper configurable

### PR DESCRIPTION
**Pull request by jul-sh**
One can now specify which kernel binary should be wrapped using feature flags.

### CODE REVIEW

File: oak_restricted_kernel_wrapper/build.rs [hunk 1]
```diff
@@ -24,7 +24,18 @@ fn main() {
...
+        (false, true) => "oak_restricted_kernel_simple_io_bin",
+        (true, true) => panic!("oak_restricted_kernel_simple_io_bin and feature oak_restricted_kernel_bin cannot be enabled at the same time. Only either version can be built."),
+        (false, false) => panic!("One of oak_restricted_kernel_simple_io_bin and feature oak_restricted_kernel_bin must be enabled.")
+    };
+
+    if cfg!(feature = "oak_restricted_kernel_simple_io_bin") {
+        "oak_restricted_kernel_simple_io_bin"
+    } else {
+        "oak_restricted_kernel_bin"
+    };
```
Comment (jul-sh) [inline oak_restricted_kernel_wrapper/build.rs:38]: Is this pretty? No. But it gets the job done. 

I don't like that we're repurposing rust features to set which binary to wrap, but it's the most pragmatic way to pass args to the build script. 

We could instead also simply rely on the caller of this crate overwriting the input file with the right version in advance, but it's imo messier.

File: oak_restricted_kernel_wrapper/build.rs [hunk 2]
```diff
@@ -24,7 +24,18 @@ fn main() {
     println!("cargo:rerun-if-changed=layout.ld");
     println!("cargo:rustc-link-arg=--script=layout.ld");
     let kernel_directory = "oak_restricted_kernel_bin";
-    let file_name = "oak_restricted_kernel_simple_io_bin";
+    let file_name = match (cfg!(feature = "oak_restricted_kernel_bin"), cfg!(feature = "oak_restricted_kernel_simple_io_bin")) {
+        (true, false) => "oak_restricted_kernel_bin",
+        (false, true) => "oak_restricted_kernel_simple_io_bin",
+        (true, true) => panic!("oak_restricted_kernel_simple_io_bin and feature oak_restricted_kernel_bin cannot be enabled at the same time. Only either version can be built."),
```
Comment (andrisaar) [inline oak_restricted_kernel_wrapper/build.rs:30]: "Feature oak_restricted[...]" would be slightly more readable and mirror the other panic error message.

**Review by andrisaar — APPROVED** [review]

**Review by tiziano88 — COMMENTED** [review]
uhmmmm I'm not a fan of this use of features. features in Rust are supposed to be additive. Also, this means that switching features overrides the binary with the same name, which is confusing (and will confuse transparent release related stuff at some point).

How about defining additional [bin] sections in that crate, and do the switching in the main?

**Comment by tiziano88** [comment]
Maybe what I'm suggesting is not so easy because the switching logic must live in build.rs though, which may not be an option for bins inside a crate.

**Comment by jul-sh** [comment]
> this means that switching features overrides the binary with the same name

how so? The output file varies also. Note that this crate is special, and should not be built using `cargo build` directly https://github.com/project-oak/oak/blob/main/oak_restricted_kernel_wrapper/README.md/#L11

> Maybe what I'm suggesting is not so easy because the switching logic must live in build.rs though, which may not be an option for bins inside a crate.

yup, that's the difficult part.

**Comment by tiziano88** [comment]
we should think about how to do this in bazel too eventually (@ernoc FYI)

**Review by tiziano88 — APPROVED** [review]

**Comment by conradgrobler** [comment]
> Maybe what I'm suggesting is not so easy because the switching logic must live in build.rs though, which may not be an option for bins inside a crate.

The only two options I can think of are not great either:

- Pass some environment variable as part of the `just` command and read that in the build script
- Always copy both versions of the restricgted kernel to two locations and then include each one in each of the binaries

**Review by conradgrobler — APPROVED** [review]

**Pull request merged**